### PR TITLE
Merge master into formplayer

### PR DESCRIPTION
**Pull request by orangejenny**
There are several case claim improvements not yet available to formplayer, notably https://github.com/dimagi/commcare-core/pull/959 which is needed soon for the new USH dedup workflow.

I'm hoping to deploy this the week of the 4th and also deploy https://github.com/dimagi/formplayer/pull/787 as per Shubham's [comment](https://github.com/dimagi/formplayer/pull/787#issuecomment-743177409).

@shubham1g5 @ShivamPokhriyal @calellowitz @ctsims could you each look over your commits and approve this?

I am not currently planning to QA this but am open to disagreement.

### CODE REVIEW

**Review by shubham1g5 — COMMENTED** [review]
Looks good to me. There are some minor changes this PR will require on Formplayer side as seen in this [commit](https://github.com/dimagi/formplayer/pull/799/commits/7011ed5ac9075fb07bd832dc3feac1ccd9067475). I am happy to PR them separately with just the changes needed to support the default value in case claim if needed. 

I think it would be nice for this to go through a cursory QA given the large amount of changes. 

The bigger challenge in getting this out is on infra side though as bumping the DB version would force fresh restores for all users. @ctsims  and @dannyroberts  were looking at this issue earlier and we should wait for a go ahead from one of them before we deploy this.

**Comment by ctsims** [comment]
@orangejenny I 100% think we should QA this.

@shubham1g5 I will read through to review for model changes, but is there a particular one that you think will require a userdb rebuild? There are 100% changes that will require app rebuilds, but I did a cursory review the other day and didn't spot any userdb changes specifically.

File: src/main/java/org/commcare/session/RemoteQuerySessionManager.java [hunk 1]
```diff
@@ -40,6 +45,16 @@ private RemoteQuerySessionManager(RemoteQueryDatum queryDatum,
                                       EvaluationContext evaluationContext) {
         this.queryDatum = queryDatum;
         this.evaluationContext = evaluationContext;
+        initUserAnswers();
+    }
+
+    private void initUserAnswers() {
+        OrderedHashtable<String, QueryPrompt> queryPrompts = queryDatum.getUserQueryPrompts();
+        for (Enumeration en = queryPrompts.keys(); en.hasMoreElements(); ) {
+            String promptId = (String)en.nextElement();
```
Comment (ctsims) [inline src/main/java/org/commcare/session/RemoteQuerySessionManager.java:54]: I think this needs to get updated to make sure this pr is included:
https://github.com/dimagi/commcare-core/pull/959

which has the init behavior for the query screen that USH needs for performing de-dupe.

File: src/main/java/org/javarosa/core/model/FormDef.java [hunk 2]
```diff
@@ -1343,177 +1347,16 @@ public String fillTemplateString(String template, TreeReference contextRef, Hash
...
-     * Identify the itemset in the backend model, and create a set of SelectChoice
-     * objects at the current question reference based on the data in the model.
-     *
-     * Will modify the itemset binding to contain the relevant choices
-     *
-     * @param itemset The binding for an itemset, where the choices will be populated
-     * @param curQRef A reference to the current question's element, which will be
-     *                used to determine the values to be chosen from.
-     */
-    public void populateDynamicChoices(ItemsetBinding itemset, TreeReference curQRef) {
```
Comment (ctsims) [inline src/main/java/org/javarosa/core/model/FormDef.java:1356]: @shubham1g5 Shout out again for refactoring this code out. Anything that drops the size of the FormDef object by 200 lines is a huge improvement to the readability of the code.

**Review by ctsims — APPROVED** [review]
Only one concrete thing I think needs to be updated. 

I didn't identify anything in this PR that would require a UserDB rebuild, by the way, so I think we can avoid a cold start. Would be very helpful for someone to confirm, though.

File: build.gradle [hunk 3]
```diff
@@ -70,7 +70,7 @@ dependencies {
     compile 'joda-time:joda-time:2.9.4'
     compile 'com.carrotsearch:hppc:0.7.2'
     compile 'com.squareup.retrofit2:retrofit:2.3.0'
-    compile 'com.squareup.okhttp3:okhttp:4.4.1'
+    compile 'com.squareup.okhttp3:okhttp:3.12.12'
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply build.gradle:73]: @ShivamPokhriyal

**Comment by orangejenny** [comment]
Cool, [QA ticket](https://dimagi-dev.atlassian.net/browse/QA-2218) created. I just asked for a formplayer regression and verified that includes case claim. If there's anything additional QA should focus on, please comment on that ticket.

@shubham1g5 So https://github.com/dimagi/formplayer/pull/799, or at least that commit, should be deployed at the same time as this PR? That sounds fine.

I'm planning to chat with @dannyroberts about the rollout, yes. Even if it's only app rebuilds, we might want to consider deploying over two nights instead of all at once.

File: src/main/java/org/commcare/session/RemoteQuerySessionManager.java (hunk 1 above)
Comment (orangejenny) [reply src/main/java/org/commcare/session/RemoteQuerySessionManager.java:55]: Thanks for flagging, my master was a couple of commits behind HEAD. That PR is now included, along with a few more commits from @shubham1g5 and @stephherbers

Comment (ctsims) [reply src/main/java/org/commcare/session/RemoteQuerySessionManager.java:55]: great, thanks. Reviewed the new commits and didn't see any UserDB changes in those either.

File: build.gradle (hunk 3 above)
[the remark this answers is not in the document]
Comment (shubham1g5) [reply build.gradle:73]: This was intentionally done for mobile as newer okhttp versions only provide support for Android 5+ and `3.12.12` is a okhttp backport for apps supporting Android 4 or lesser.

[the remark this answers is not in the document]
Comment (ctsims) [reply build.gradle:73]: Ah, thanks for confirming. I imagine this should be fine for formplayer and if there are any issues they'd show up in a general QA pass.

**Comment by shubham1g5** [comment]
> I didn't identify anything in this PR that would require a UserDB rebuild, by the way, so I think we can avoid a cold start. Would be very helpful for someone to confirm, though.

Went through commits and confirming this is the case indeed. This merge seem harmless in that scenario though we won't be able to push https://github.com/dimagi/formplayer/pull/787  out with this. 

> So dimagi/formplayer#799, or at least that commit, should be deployed at the same time as this PR? That sounds fine.

Yes, dimagi/formplayer#799 will need the corresponding HQ changes to be deployed first though while that commit can be deployed without the HQ changes.

**Comment by shubham1g5** [comment]
@orangejenny We should also update the [version here](https://github.com/dimagi/commcare-core/blob/master/src/cli/java/org/commcare/util/engine/CommCareConfigEngine.java#L71) to `50` as part of this PR as this merge will upgrade Formplayer to all changes added till CommCare 2.50 mobile version.

**Comment by orangejenny** [comment]
@shubham1g5 Thank you, just added fdc85703c to update the version.

**Comment by ctsims** [comment]
> > I didn't identify anything in this PR that would require a UserDB rebuild, by the way, so I think we can avoid a cold start. Would be very helpful for someone to confirm, though.
> 
> Went through commits and confirming this is the case indeed. This merge seem harmless in that scenario though we won't be able to push [dimagi/formplayer#787](https://github.com/dimagi/formplayer/pull/787) out with this.
> 
> > So [dimagi/formplayer#799](https://github.com/dimagi/formplayer/pull/799), or at least that commit, should be deployed at the same time as this PR? That sounds fine.
> 
> Yes, [dimagi/formplayer#799](https://github.com/dimagi/formplayer/pull/799) will need the corresponding HQ changes to be deployed first though while that commit can be deployed without the HQ changes.

Thanks for the spot check, glad to hear I didn't miss anything, and definitely happy for 787 to go out with its own lifecycle.

I think it would still be helpful for someone to physically confirm that sandboxes don't break by creating some piece of state of each type (a synced user with incomplete forms) and ensuring that all of those pieces work after updating to the new head. 

Since we don't do migrations it might actually be a good idea to have a way to easily produce a pre-saved DB file based on a CCZ and based on a restore XML file, and keep a pre-made sandbox of the "Current" clean state in the test assets. That would give us a lot more confidence into whether or not an update breaks one build or the other. I can see if I can put that together quickly to use for this.

**Comment by ctsims** [comment]
@orangejenny I made a pr in formplayer to test what I was describing https://github.com/dimagi/formplayer/pull/802 . 

I can run it against this as soon as https://github.com/dimagi/formplayer/pull/800 is updated to target this head in the cc-core library reference.

**Review by shubham1g5 — APPROVED** [review]

**Comment by orangejenny** [comment]
> I think it would still be helpful for someone to physically confirm that sandboxes don't break by creating some piece of state of each type (a synced user with incomplete forms) and ensuring that all of those pieces work after updating to the new head.

@ctsims Is this no longer needed because of https://github.com/dimagi/formplayer/pull/802? Or should I still do this?

**Comment by ctsims** [comment]
@orangejenny I think https://github.com/dimagi/formplayer/pull/802 takes care of this. I did confirm that the test caught sandboxes breaking in a few different ways.

The only thing I'm not sure about is incomplete forms, which I'd forgotten about until re-reading this. That's certainly lower risk (I'm not sure USH even uses them), but might be worth testing directly.

**Comment by orangejenny** [comment]
@ctsims Cool, I can test incomplete forms.

**Comment by orangejenny** [comment]
Confirming that I tested incomplete forms locally:
* Checked out formplayer `master` and commcare-core `formplayer`, built jar
* Opened a form and filled out some fields
* Checked out formplayer `formplayer_2.50` and commcare-core `jls/formplayer-merge-master`, built jar
* Went into incomplete forms, found form, opened it and saw previous answers, finished and submitted it.

**Comment by orangejenny** [comment]
QA is complete.

**Pull request merged**